{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Predicting disease phenotypes of <i>de novo</i> ATP1A3 variants\n",
    "## Biren Dave\n",
    "## Generating PROVEAN and SIFT scores for variants\n",
    "\n",
    "The following code cell reads the spreadsheet containing the ATP1A3 mutations and generates an output file compatible with PROVEAN. `file_reader` reads an input `delim`-delimited file and generated an ordered dictionary. \n",
    "\n",
    "PROVEAN requires input data to be space-delimited and formatted like this: `<Protein_ID> <position> <reference_residue(s)> <variant_residue(s)> <comment (optional)>`. Multiple variants can be analyzed at once if they are line-delimited. To achieve appropriate formatting, the `provean_input_formatter` function has been implemented. Finally, an output `.txt` file is generated by the function `output_writer`. This file can be directly uploaded to the online [PROVEAN server](http://provean.jcvi.org/protein_batch_submit.php?species=human). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "\n",
    "NCBI_ref_id = \"NP_689509.1\"\n",
    "\n",
    "def file_reader(infile, delim):\n",
    "    data = []\n",
    "    with open(infile, \"r\") as fh:\n",
    "        for row in csv.DictReader(fh, delimiter = delim):\n",
    "            data.append(row)\n",
    "    return data\n",
    "\n",
    "def provean_input_formatter(l):\n",
    "    provean_input = []\n",
    "    for variant in l:\n",
    "        temp = []\n",
    "        if variant[\"type\"] == \"Missense_Mutation\":\n",
    "            ref_aa = variant[\"substitution\"][0]\n",
    "            alt_aa = variant[\"substitution\"][-1]\n",
    "            pos = variant[\"substitution\"][1:-1]\n",
    "            disease = variant[\"disease\"]\n",
    "            temp = [NCBI_ref_id, pos, ref_aa, alt_aa, disease]\n",
    "            provean_input.append(temp)\n",
    "        else:\n",
    "            pass\n",
    "    return provean_input\n",
    "\n",
    "def output_writer(l, outfile, delim):\n",
    "    with open(outfile, \"w\") as fh:\n",
    "        writer = csv.writer(fh, delimiter = delim)\n",
    "        writer.writerows(l)\n",
    "            \n",
    "variants = file_reader(\"mutations.csv\", \",\")\n",
    "provean_input = provean_input_formatter(variants)\n",
    "output_writer(provean_input, \"provean_input.txt\", \" \")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Making a feature matrix \n",
    "\n",
    "The tabulated PROVEAN output is then related back to the originial mutation data, so that disease outcome and protein domain data can be added. The dictonary `domains` maps certain amino acid positions to NCBI-annotated protein regions, `properties` maps single letter amino acid codes to a set of chemical properties, and `diseases` maps amino acid changes (in the format A[pos]B, where A is the reference residue and B is the altered residue) to disease phenotypes according to published data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "NCBI_domains = {37: \"phosphorylation site\",\n",
    "           56: \"phosphorylation site\",\n",
    "           range(72,75): \"interaction with phosphoinositide-3 kinase\",\n",
    "           range(78,99): \"transmembrane region\",\n",
    "           range(122,143): \"transmembrane region\",\n",
    "           218: \"phosphorylation site\",\n",
    "           265: \"phosphorylation site\",\n",
    "           range(279,299): \"transmembrane region\",\n",
    "           range(311,329): \"transmembrane region\",\n",
    "           442: \"phosphorylation site\",\n",
    "           548: \"phosphorylation site\",\n",
    "           range(763,783): \"transmembrane region\",\n",
    "           range(834,857): \"transmembrane region\",\n",
    "           range(909,929): \"transmembrane region\",\n",
    "           933: \"phosphorylation site by PKA\",\n",
    "           range(942,961): \"transmembrane region\",\n",
    "           range(976,997): \"transmembrane region\"}\n",
    "'''\n",
    "\n",
    "pfam_domains = {range(33,102): \"cation_ATPase_N\",\n",
    "                range(153,345): \"E1-E2_ATPase\",\n",
    "                range(416,512): \"cation_ATPase\",\n",
    "                range(789,999): \"cation_ATPase_C\"}\n",
    "\n",
    "properties = {\"R\": \"charged\",\n",
    "              \"K\": \"charged\",\n",
    "              \"D\": \"charged\",\n",
    "              \"E\": \"charged\",\n",
    "              \"Q\": \"polar\",\n",
    "              \"N\": \"polar\",\n",
    "              \"H\": \"polar\",\n",
    "              \"S\": \"polar\",\n",
    "              \"T\": \"polar\",\n",
    "              \"Y\": \"polar\",\n",
    "              \"C\": \"polar\",\n",
    "              \"W\": \"polar\",\n",
    "              \"A\": \"hydrophobic\",\n",
    "              \"I\": \"hydrophobic\",\n",
    "              \"L\": \"hydrophobic\",\n",
    "              \"M\": \"hydrophobic\",\n",
    "              \"F\": \"hydrophobic\",\n",
    "              \"V\": \"hydrophobic\",\n",
    "              \"P\": \"hydrophobic\",\n",
    "              \"G\": \"hydrophobic\"}\n",
    "\n",
    "def get_domain(position):\n",
    "    for key in pfam_domains.keys():\n",
    "        if position in key:\n",
    "            return pfam_domains[key]\n",
    "    return \"unannotated domain\"\n",
    "\n",
    "def get_chemical_properties(amino_acid):\n",
    "    return properties[amino_acid]\n",
    "\n",
    "def prop_features(prop):\n",
    "    if prop == \"charged\":\n",
    "        return [1, 0, 0]\n",
    "    elif prop == \"polar\":\n",
    "        return [0, 1, 0]\n",
    "    elif prop == \"hydrophobic\":\n",
    "        return [0, 0, 1]\n",
    "\n",
    "def ncbi_domain_features(domain):\n",
    "    if domain == \"phosphorylation site\":\n",
    "        return [1, 0, 0, 0]\n",
    "    elif domain == \"interaction with phosphoinositide-3 kinase\":\n",
    "        return [0, 1, 0, 0]\n",
    "    elif domain == \"transmembrane region\":\n",
    "        return [0, 0, 1, 0]\n",
    "    elif domain == \"phosphorylation site by PKA\":\n",
    "        return [0, 0, 0, 1]\n",
    "\n",
    "def pfam_domain_features(domain):\n",
    "    if domain == \"cation_ATPase_N\":\n",
    "        return [1, 0, 0, 0, 0]\n",
    "    elif domain == \"E1-E2_ATPase\":\n",
    "        return [0, 1, 0, 0, 0]\n",
    "    elif domain == \"cation_ATPase\":\n",
    "        return [0, 0, 1, 0, 0]\n",
    "    elif domain == \"cation_ATPase_C\":\n",
    "        return [0, 0, 0, 1, 0]\n",
    "    elif domain == \"unannotated domain\":\n",
    "        return [0, 0, 0, 0, 1]\n",
    "    \n",
    "\n",
    "def provean_sift_features(provean, sift):\n",
    "    if provean < -2.5:\n",
    "        p_pred = 1\n",
    "    else:\n",
    "        p_pred = 0\n",
    "    if sift > 0.05:\n",
    "        s_pred = 1\n",
    "    else:\n",
    "        s_pred = 0\n",
    "    return [provean, p_pred, sift, s_pred]\n",
    "    \n",
    "diseases = {}\n",
    "\n",
    "for variant in variants:\n",
    "    diseases[variant[\"substitution\"].strip()] = variant[\"disease\"]\n",
    "\n",
    "provean_output = file_reader(\"provean_output.tsv\", \"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, I define the custom object `Variant`, which contains a method `get_feature_vector` which constructs a 16-dimensional feature vector for that variant. These feature vectors are stored in the list `feature_vectors`, and their corresponding disease state is stored in the list `disease_list`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Variant():\n",
    "    def __init__(self, ref_aa, pos, alt_aa, provean, sift):\n",
    "        self.ref_aa = ref_aa\n",
    "        self.alt_aa = alt_aa\n",
    "        self.pos = pos\n",
    "        self.domain = get_domain(pos)\n",
    "        #self.domain_features = pfam_domain_features(self.domain)\n",
    "        self.ref_aa_prop = get_chemical_properties(ref_aa)\n",
    "        self.alt_aa_prop = get_chemical_properties(alt_aa)\n",
    "        self.provean = provean\n",
    "        self.sift = sift\n",
    "        try:\n",
    "            self.disease = diseases[ref_aa + str(pos) + alt_aa]\n",
    "        except KeyError:\n",
    "            self.disase = \"null\"\n",
    "    def get_feature_vector(self):\n",
    "        aa_prop_vector = prop_features(self.ref_aa_prop) + prop_features(self.alt_aa_prop)\n",
    "        domain_vector = pfam_domain_features(self.domain)\n",
    "        provean_sift_vector = provean_sift_features(self.provean, self.sift)\n",
    "        return [self.pos] + aa_prop_vector + domain_vector + provean_sift_vector \n",
    "    def get_disease(self):\n",
    "        return self.disease\n",
    "    def __str__(self):\n",
    "        return \"ref_aa: {}, alt_aa: {}, pos: {}, ref_aa_prop: {}, alt_aa_prop: {}, PROVEAN_score: {}, SIFT_score: {}\".format(\n",
    "                self.ref_aa, self.alt_aa, self.pos, self.ref_aa_prop, self.alt_aa_prop, self.provean, self.sift)\n",
    "\n",
    "feature_vectors = []\n",
    "disease_list = []\n",
    "plot_data = []\n",
    "\n",
    "for var in provean_output:\n",
    "    if var[\"PROTEIN_ID\"] == NCBI_ref_id:\n",
    "        v = Variant(var[\"RESIDUE_REF\"], int(var[\"POSITION\"]), var[\"RESIDUE_ALT\"], float(var[\"pSCORE\"]), float(var[\"sSCORE\"]))\n",
    "        feature_vectors.append(v.get_feature_vector())\n",
    "        disease_list.append(v.get_disease())\n",
    "        \n",
    "        plot_data.append([v.pos, v.provean, v.sift, v.get_disease()])\n",
    "\n",
    "with open(\"ATP1A3_variants_2D.csv\", \"w\") as fh:\n",
    "    writer = csv.writer(fh, delimiter = \",\")\n",
    "    writer.writerows(plot_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Training the Support Vector Classifier (SVC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,\n",
       "  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',\n",
       "  max_iter=-1, probability=False, random_state=None, shrinking=True,\n",
       "  tol=0.001, verbose=False)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn import svm\n",
    "\n",
    "classifier = svm.SVC()\n",
    "classifier.fit(feature_vectors, disease_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Making predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "pos = input(\"enter position of the amino acid variant: \")\n",
    "ref_aa = input(\"enter the single letter code of the reference amino acid: \")\n",
    "alt_aa = input(\"enter the single letter code of the altered amino acid: \")\n",
    "provean = input(\"enter the PROVEAN score of the variant: \")\n",
    "sift = input(\"enter the SIFT score of the variant: \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "predicted disease phenotype: CAPOS\n"
     ]
    }
   ],
   "source": [
    "v = Variant(ref_aa, int(pos), alt_aa, float(provean), float(sift))\n",
    "\n",
    "print(\"predicted disease phenotype: \" + classifier.predict([v.get_feature_vector()])[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9814814814814815"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "classifier.score(feature_vectors, disease_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
